### PR TITLE
LevelZero NPU support + DBK support

### PR DESCRIPTION
**Pull request by franz**

### CODE REVIEW

**Comment by linehill** [comment]
Can we have a CI to test PoCL building with -DENABLE_NPU=ON to avoid bit-rotting?

```
mnt/md1/linehill/ws-pocl-2/pocl/lib/CL/devices/level0/npu_dbk_matmul.cc:86:3: error: ‘cl_tensor_layout_ml’ was not declared in this scope; did you mean ‘cl_tensor_layout_ml_exp’?
   86 |   cl_tensor_layout_ml *L = nullptr;
      |   ^~~~~~~~~~~~~~~~~~~
      |   cl_tensor_layout_ml_exp
/mnt/md1/linehill/ws-pocl-2/pocl/lib/CL/devices/level0/npu_dbk_matmul.cc:86:24: error: ‘L’ was not declared in this scope
   86 |   cl_tensor_layout_ml *L = nullptr;
      |                        ^
/mnt/md1/linehill/ws-pocl-2/pocl/lib/CL/devices/level0/npu_dbk_matmul.cc:88:9: error: ‘cl_dbk_attributes_exp_matmul’ does not name a type; did you mean ‘cl_dbk_attributes_matmul_exp’?
   88 |   const cl_dbk_attributes_exp_matmul *Attrs
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         cl_dbk_attributes_matmul_exp
/mnt/md1/linehill/ws-pocl-2/pocl/lib/CL/devices/level0/npu_dbk_matmul.cc:90:42: error: ‘Attrs’ was not declared in this scope
   90 |   ReplaceMap["SHAPE_M"] = std::to_string(Attrs->a.shape[0]);
      |                                          ^~~~~
/mnt/md1/linehill/ws-pocl-2/pocl/lib/CL/devices/level0/npu_dbk_matmul.cc:104:29: error: expected primary-expression before ‘)’ token
  104 |   L = (cl_tensor_layout_ml *)Attrs->a.layout;
      |                             ^
/mnt/md1/linehill/ws-pocl-2/pocl/lib/CL/devices/level0/npu_dbk_matmul.cc:106:29: error: expected primary-expression before ‘)’ token
  106 |   L = (cl_tensor_layout_ml *)Attrs->b.layout;
      |                             ^
/mnt/md1/linehill/ws-pocl-2/pocl/lib/CL/devices/level0/npu_dbk_matmul.cc:110:29: error: expected primary-expression before ‘)’ token
  110 |   L = (cl_tensor_layout_ml *)Attrs->c.layout;
      |                             ^
(...)
```

File: lib/CL/devices/level0/level0-compilation.cc [hunk 1]
```diff
@@ -1287,3 +2118,93 @@ bool Level0CompilationJobScheduler::getBestKernel(Level0Program *Prog,
   }
   return Res;
 }
+
+#ifdef ENABLE_NPU
+Level0BuiltinProgram *Level0CompilationJobScheduler::createBuiltinProgram(
+    ze_context_handle_t Ctx,
```
Comment (pjaaskel) [inline lib/CL/devices/level0/level0-compilation.cc:2124]: Indent.

File: lib/CL/devices/level0/level0-driver.cc [hunk 2]
```diff
@@ -3453,17 +3591,119 @@ int Level0Device::createProgram(cl_program Program, cl_uint DeviceI) {
...
+  pocl_append_to_buildlog(Program, DeviceI,
+                   strdup(BuildLog.c_str()),
+                   BuildLog.size());
+  return CL_BUILD_PROGRAM_FAILURE;
+#endif
+}
+
+
 int Level0Device::freeProgram(cl_program Program, cl_uint DeviceI) {
+  /* module can be NULL if compilation fails */
```
Comment (pjaaskel) [inline lib/CL/devices/level0/level0-driver.cc:3640]: //

File: lib/CL/devices/level0/npu/include/ze_command_queue_npu_ext.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
+/*
```
Comment (pjaaskel) [inline lib/CL/devices/level0/npu/include/ze_command_queue_npu_ext.h:1]: Why do we copy these headers? Please add a link to the source somewhere?

File: lib/CL/devices/level0/pocl-level0.cc [hunk 4]
```diff
@@ -678,6 +819,13 @@ int pocl_level0_link_program(cl_program Program, cl_uint DeviceI,
                              int CreateLibrary) {
   cl_device_id Dev = Program->devices[DeviceI];
   Level0Device *Device = (Level0Device *)Dev->data;
+
+  if (Dev->compiler_available == CL_FALSE ||
```
Comment (pjaaskel) [inline lib/CL/devices/level0/pocl-level0.cc:823]: I think we can assume ! works for CL_FALSE, let's use that, more readable.

File: lib/CL/devices/level0/pocl-level0.cc [hunk 5]
```diff
@@ -897,34 +1038,144 @@ int pocl_level0_setup_metadata(cl_device_id Device, cl_program Program,
...
+int
+pocl_npu_validate_khr_gemm (cl_bool TransA,
+                            cl_bool TransB,
+                            const cl_tensor_desc_exp *TenA,
+                            const cl_tensor_desc_exp *TenB,
+                            const cl_tensor_desc_exp *TenCIOpt,
+                            const cl_tensor_desc_exp *TenCOut,
+                            const cl_tensor_datatype_value_exp *Alpha,
+                            const cl_tensor_datatype_value_exp *Beta)
+{
```
Comment (pjaaskel) [inline lib/CL/devices/level0/pocl-level0.cc:1065]: Hm. This seems to be a .cc source. I wonder why formatter treats it as C?

File: lib/CL/devices/level0/level0-compilation.cc (hunk 1 above)
Comment (linehill) [reply lib/CL/devices/level0/level0-compilation.cc]: Should be fixed not (ran format-branch).

File: lib/CL/devices/level0/level0-driver.cc (hunk 2 above)
Comment (linehill) [reply lib/CL/devices/level0/level0-driver.cc]: Should be fixed now.

File: lib/CL/devices/level0/pocl-level0.cc (hunk 4 above)
Comment (linehill) [reply lib/CL/devices/level0/pocl-level0.cc]: Done.

File: lib/CL/devices/level0/pocl-level0.cc (hunk 5 above)
Comment (linehill) [reply lib/CL/devices/level0/pocl-level0.cc]: Should be fixed now (ran format-branch).

File: lib/CL/devices/level0/npu/include/ze_command_queue_npu_ext.h (hunk 3 above)
Comment (linehill) [reply lib/CL/devices/level0/npu/include/ze_command_queue_npu_ext.h:3]: I think these are not distributed via the base level0 package. Added link to the origin (I think they are from) into the copied headers.

**Comment by linehill** [comment]
> Also please add some brief release notes ...

Done.

File: lib/CL/devices/level0/npu/include/ze_command_queue_npu_ext.h (hunk 3 above)
Comment (franz) [reply lib/CL/devices/level0/npu/include/ze_command_queue_npu_ext.h:3]: not distributed in base, and also AFAICT there are no github or debian packages.

**Pull request merged**